### ISSUE TITLE
Extended lambdas bypass positivity checking in records

### ISSUE THREAD

**Issue by Saizan** [agda/agda#2963]
Agda does not complain about the following
```agda
  record ⊤ : Set where
    constructor tt

  unit-elim : ∀ {ℓ} {A : Set ℓ} → ⊤ → (⊤ → A) → A
  unit-elim u t = t u

  record R : Set where
    coinductive
    field
      f : ∀ u → unit-elim u (\ { _ → R → R })
```

It does complain if you try the same with "data".

**Comment by andreasabel** [comment]
Here is the subsequent proof of False:
```agda
data ⊥ : Set where
record ⊤ : Set where

record R : Set where
  -- coinductive  -- Not needed, since not detected as recursive.
  field
    app : let
        z : ⊤ → Set
        z = λ{ _ → R → ⊥ }
      in z record{}
open R

abs : (R → ⊥) → R
abs g .app x = g x

delta : R
delta = abs λ x → app x x

loop : ⊥
loop = app delta delta
```

**Comment by nad** [comment]
Agda 2.4.2 rejects the first piece of code above, but Agda 2.4.2.1 accepts it (if `coinductive` is removed).

**Comment by nad** [comment]
Bisection points to 43803ccd077d4f033e604fcb5fe6c10cb1d87164: "[ refactor ] Factored out scopeCheckExtendedLambda".

**Comment by andreasabel** [comment]
I can see why __this testcase__ would work ok before 43803cc but not after.
Because in this commit I removed the call to `reduce` that would inline the extended lambda.

However, what about the following variant, where the extended lambda cannot be reduced away?
```agda
data Bool : Set where
  true false : Bool

data ⊥ : Set where
record ⊤ : Set where

record R (b : Bool) : Set where
  constructor abs
  field
    app : let
        z : Bool → Set
        z = λ{ true → R b → ⊥; false → R b → ⊥ }
      in z b
open R

delta : R true
delta = abs λ x → app x x

loop : ⊥
loop = app delta delta
```
My oldest Agda, 2.4.0 accepts this proof of false.
I suspect that positivity checker never traversed into extended lambda bodies.
@nad, do you have older Agdas than 2.4.0 on which you could test it?

**Comment by andreasabel** [comment]
For a comparable data type, the positivity checker is invoked on a set of names including the extended lambda:
```
positivity of [D, .Issue2963-Bool..extendedlambda0, abs]
```
Yet for `R`, this set is just the name of the record type:
```
positivity of [R]
```
The bug is thus not in `TypeChecking.Positivity`, but in the general declaration checking, which calls the positivity checker.

**Comment by andreasabel** [comment]
I think I found the offending commit :P  e7f4891a85a822fe900b75f2d5b42733812cce5b
```
Record declarations are no longer seen as mutual blocks.
```
It introduced a special case for record decls here:
https://github.com/agda/agda/blob/e7f4891a85a822fe900b75f2d5b42733812cce5b/src/full/Agda/Syntax/Concrete/Definitions.hs#L348-L354
Removing this optimization fixes the positivity issue.

**Comment by nad** [comment]
> @nad, do you have older Agdas than 2.4.0 on which you could test it?

Rejected by 2.3.0.1, accepted by 2.3.2.

**Comment by nad** [comment]
> Removing this optimization fixes the positivity issue.

I seem to recall that this optimisation (implemented in Fischbachau) made Agda considerably faster for some code. Hopefully the optimisation is no longer needed.

**Comment by andreasabel** [comment]
I hope so, too.  Do you remember why the mutual block caused the bad performance then?

**Comment by nad** [comment]
No, I don't. This bug suggests that it had something to do with positivity checking, and positivity checking has been optimised several times since then.

**Comment by UlfNorell** [comment]
There is a nontrivial performance degradation. On my machine running `library-test` spends `2.6s` on termination checking before 553b281, and `6.0s` after. The effect on positivity checking is less noticable: `3.2s` to `4.8s`.

Of course, this isn't a huge problem (the entire run takes about `2m30s`), but there might be cases where the problem is more severe.

**Comment by andreasabel** [comment]
What causes the performance degradation?  Are termination and positivity checks run twice now for records?

**Comment by UlfNorell** [comment]
I'm not sure, I didn't investigate. I just spent the afternoon thinking I caused it, and was relieved to see bisection pointing here, so I thought I'd share the numbers.

**Comment by nad** [comment]
Perhaps it would make sense to implement a safe version of the old optimisation. Would it be safe to restrict the mutual block to the underlying record type, including definitions that precede some field and definitions inside field declarations (like anonymous lambdas)?

**Comment by nad** [comment]
I found the following code, which might be relevant to the discussion above:

https://github.com/agda/agda/blob/87fb8278650472a3c9d93badd7888ced15b51240/src/full/Agda/TypeChecking/Rules/Decl.hs#L435-L440

**Comment by Saizan** [comment]
Also this bit of code could be relevant: an opportunity to run checks on extended lambda definitions outside a mutual block. Though if they are indeed mutually defined with something else it doesn't help I suppose.

https://github.com/agda/agda/blob/87fb8278650472a3c9d93badd7888ced15b51240/src/full/Agda/TypeChecking/Rules/Term.hs#L631-L633

**Comment by nad** [comment]
Consider the following code in `checkDecl`:

https://github.com/agda/agda/blob/87fb8278650472a3c9d93badd7888ced15b51240/src/full/Agda/TypeChecking/Rules/Decl.hs#L173

https://github.com/agda/agda/blob/87fb8278650472a3c9d93badd7888ced15b51240/src/full/Agda/TypeChecking/Rules/Decl.hs#L182

https://github.com/agda/agda/blob/87fb8278650472a3c9d93badd7888ced15b51240/src/full/Agda/TypeChecking/Rules/Decl.hs#L190

Why do we run `mutualChecks` twice for record type definitions? Now that all record modules are treated as being in mutual blocks I guess that it is safe to remove `mutual empty [d] $`. This might make Agda a bit faster in some cases.

**Comment by nad** [comment]
I missed that the returned checks are skipped inside mutual blocks, so it should be safe to remove `mutual empty [d]` (or rather, replace it with `impossible`), but this should not have any effect.

**Comment by nad** [comment]
> I found the following code, which might be relevant to the discussion above:

I don't think this special case could be triggered any longer. I removed it.

> What causes the performance degradation? Are termination and positivity checks run twice now for records?

The termination checker's performance degradation might, in part, be related to the quadratic behaviour that I just fixed (#3590). I did not benchmark the code properly, because the standard library test is currently broken.
